### DISCUSSION TITLE
Suspense for mutation

### DISCUSSION THREAD

**Discussion by DamodarSojka** [TanStack/query#967] [General]
Is there a way to trigger suspense mode for useMutation hook?

**Comment by dburles** [comment]
Just ran into this myself. It seems as though `useMutation` does not suspend. Though, I did see @tannerlinsley mention that it should over here: https://github.com/tannerlinsley/react-query/discussions/344#discussioncomment-4167. Can anyone shed any light on this?

**Reply by dburles** [reply]
Still curious if anyone can answer this!

**Comment by yashino91** [comment]
Hello! I am using react-query 3.19.1. In our new application we'd like to handle any loading state with `React.Suspense` and this is working fine with `useQuery`. But it seems that `useMutation` still does not suspend. Is there any information on how to achive it or is it not intended to work that way?

**Comment by TkDodo** [comment]
looking at the code, there is support for `useErrorBoundary`, but not for `suspense`

**Comment by hedaukartik** [comment]
Why not suspense for useMutation?

**Reply by TkDodo** [reply]
because no on has implemented it I guess 😅

**Comment by SennaSanzo** [comment]
With the release of React 19, is using the ```use()``` function together with ```your_mutation.mutateAsync(...)``` a valid way to trigger Suspense?

**Reply by TkDodo** [reply]
no because you would usually trigger that in an event handler and calling `use` in an event handler does nothing.

**Reply by SennaSanzo** [reply]
Is it too hacky to store the mutateAsync Promise in state and call use() on it during a re-render? 😂
Joke aside, real question. I'm pretty new in the React community. It works but doesn't feel right so...

**Comment by lizzyTheLizard** [comment]
FYI: I solved the problem on my side using the following to generate a "suspense aware mutation". It most likely will not work for all cases, but it seems to works for mine...

```
function useSuspenseMutation<TData = unknown, TError = DefaultError, TVariables = void, TOnMutateResult = unknown>(options: UseMutationOptions<TData, TError, TVariables, TOnMutateResult>, queryClient?: QueryClient): UseMutationResult<TData, TError, TVariables, TOnMutateResult> {
  // This is a hack as tanstack/react-query does not provide a useSuspenseMutation hook. It will not work perfectly in all cases.
  const [endSuspenseFunction, setEndSuspenseFunction] = useState<(() => void) | undefined>(undefined)

  const onSettled: typeof options.onSettled = (d, e, v, o, c) => {
    endSuspenseFunction?.()
    return options.onSettled?.(d, e, v, o, c)
  }

  const mutation = useMutation({ ...options, onSettled }, queryClient)
  if (mutation.isPending) {
    // eslint-disable-next-line @typescript-eslint/only-throw-error
    throw new Promise<void>((resolve) => {
      setEndSuspenseFunction(() => { resolve() })
    })
  }
  return mutation
}
```
